pystorm/spout.py

```diff
@@ -97,3 +91,76 @@ def _run(self):
         else:
             self.logger.error('Received invalid command from Storm: %r', cmd)
         self.send_message({'command': 'sync'})
+
+
+class ReliableSpout(Component):
+    """Reliable spout that will automatically replay failed tuples.
+
+    Failed tuples will be replayed up to ``max_fails`` times.
+
+    For more information on spouts, consult Storm's
+    `Concepts documentation <http://storm.apache.org/documentation/Concepts.html>`_.
+    """
+    max_fails = 3
+
+    def __init__(self, *args, **kwargs):
+        super(ReliableSpout, self).__init__(*args, **kwargs)
+        self.failed_tuples = Counter()
+        self.unacked_tuples = {}
+
+    def ack(self, tup_id):
+        """Called when a bolt acknowledges a Tuple in the topology.
+
+        :param tup_id: the ID of the Tuple that has been fully acknowledged in
+                       the topology.
+        :type tup_id: str
+        """
+        if tup_id in self.failed_tuples:
+            del self.failed_tuples[tup_id]
+        if tup_id in self.unacked_tuples:
+            del self.unacked_tuples[tup_id]
```

Comment:
I changed to popping from `failed_tuples` and doing a `try/except` for deleting from `unacked_tuples` so that we're not computing the hash of `tup_id` more than once.